### PR TITLE
Support for Ensemble of Unitary trajectories and Two-Time Correlations

### PR DESCRIPTION
**Pull request by Gauthameshwar**
### 1. MPS Autocorrelators helpers
Implemented `apply_observable_inplace()` and `mixed_expectation()` that is used to compute the autocorrelation values at each timepoints in the trajectory. The current changes also supports arbitrary next-nearest neighbor observables in periodic systems, which was broken previously because it couldn't handle 0 <-> L-1 term. 

### 2. Parallelised Unitary Ensemble
Enabled `run()` with `list[MPS]` for deterministic MPS ensemble evolution. Parallelised across ensemble members for efficient multi-trajectory simulations. 

### 3. Two-Time Correlators
Added support for both `auto_correlators` and `two_time_correlators` parameters to `AnalogSimParams`. Supports arbitrary observable pairs $(A, B)$ in computing their two-time correlation evaluate as $\langle\psi(t) | A | \phi_B(t)\rangle$ with $|\phi_B(t)\rangle = U(t) B
|\psi(0)\rangle$. It can also compute ensemble averaged two-site correlation matrices with shape `(n_pairs, n_times)` if the input is `list[MPS]`. 

### Testing
Added comprehensive tests for unitary ensemble observables and autocorrelations. Also verified periodic wrap-around behavior for periodic boundary conditions. Tested the results on a XXZ transverse-field ensemble evolution validated against ED. 

### Documentation
Made a tutorial notebook demonstrating single-state and ensemble workflows to study autocorrelation and two-time correlation function for various observables in an XXZ spin chain. Also demonstrated spin-current transport on a periodic ring and calculated its autocorrelation. 

Finally, updated CHANGELOG.md and UPGRADING.md guides.  

## Checklist

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [x] I have added migration instructions to the upgrade guide (if needed).
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

**If PR contains AI-assisted content:**

- [x] I have disclosed the use of AI tools in the PR description as per our [AI Usage Guidelines](https://github.com/munich-quantum-toolkit/yaqs/blob/main/docs/ai_usage.md).
- [x] I confirm that I have personally reviewed and understood all AI-generated content, and accept full responsibility for it.

### CODE REVIEW

File: docs/examples/ensemble_evolution.md
Comment (aaronleesander) [inline docs/examples/ensemble_evolution.md:1]: Can you shorten the name to something more generic? Like even ensemble evolution or something similar (Whatever makes sense to you)

File: docs/refs.bib
Comment (aaronleesander) [inline docs/refs.bib:1]: These should be removed. Write the citation directly in the example or wherever if needed. The refs.bib here is for our papers that contribute to YAQS

File: src/mqt/yaqs/analog/autocorrelator.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    """Undo :func:`_bubble_swaps_forward_for_wrap`."""
+    length = state.length
+    if length <= 2:
+        return
+    sw = _swap_gate_4()
+    for i in reversed(range(length - 2)):
+        _apply_two_site_nn_matrix_inplace(state, i, sw)
+
+
+def apply_observable_inplace(state: MPS, observable: Observable) -> None:
```
Comment (aaronleesander) [inline src/mqt/yaqs/analog/autocorrelator.py]: I think this function would be better as an MPS method, so MPS.apply_observable_inplace() or something similar. Look through the rest of the library and see if something else makes sense. Also for these public functions, try to shorten the names if possible.

File: src/mqt/yaqs/analog/autocorrelator.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            raise ValueError(msg)
+
+        _apply_two_site_nn_matrix_inplace(state, i, np.asarray(observable.gate.matrix, dtype=np.complex128))
+        return
+
+    msg = "Autocorrelator observable must be one-site or nearest-neighbor two-site."
+    raise ValueError(msg)
+
+
+def mixed_expectation(
```
Comment (aaronleesander) [inline src/mqt/yaqs/analog/autocorrelator.py]: This one could also benefit from being an MPS method. The main thing is I would move it out of autocorrelator.py since it is actually a very useful general function and does not need to be just tied to the autocorrelator.

File: src/mqt/yaqs/analog/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
+# Copyright (c) 2025 - 2026 Chair for Design Automation, TUM
```
Comment (aaronleesander) [inline src/mqt/yaqs/analog/ensemble.py:1]: Through all of this, can we also focus more on the "ensemble" rather than "unitary_ensemble"? Try to keep it as short and clean as possible. I think by definition this ensemble part is going to be unitary.

File: src/mqt/yaqs/core/data_structures/simulation_parameters.py [hunk 4]
```diff
@@ -285,6 +261,15 @@ def __init__(
         self.get_state = get_state
         self.show_progress = show_progress
         self.num_threads = num_threads
+        self.compute_autocorrelator = compute_autocorrelator
```
Comment (aaronleesander) [inline src/mqt/yaqs/core/data_structures/simulation_parameters.py]: I would prefer if this was less complicated. Even if for example we can define the two time correlators or auto correlator as an Observable. This can probably be made more compact by using the existing infrastructure.

File: src/mqt/yaqs/simulator.py [hunk 5]
```diff
@@ -756,13 +781,147 @@ def _run_circuit(
         _run_weak_sim(initial_state, operator, sim_params, noise_model, parallel=parallel)
 
 
+# ---------------------------------------------------------------------------
+# 13) ANALOG UNITARY ENSEMBLE (list of initial states)
+# ---------------------------------------------------------------------------
+def _run_analog_unitary_ensemble(
```
Comment (aaronleesander) [inline src/mqt/yaqs/simulator.py]: This could as well just be run_ensemble I believe. The analog is already handled by the arguments that are allowed, and maybe in the future it could be extended to digital.

File: src/mqt/yaqs/simulator.py [hunk 6]
```diff
@@ -776,12 +935,24 @@ def _run_analog(
...
-        initial_state: The initial system state as an MPS.
+        initial_state: The initial system state as an MPS, or a list of MPS states
+            for deterministic unitary analog ensemble evolution.
         operator: The Hamiltonian operator represented as an MPO.
         sim_params: Simulation parameters for analog simulation, including time step and evolution order.
         noise_model: The noise model applied during simulation.
         parallel: Flag indicating whether to run trajectories in parallel.
     """
+    # Deterministic unitary ensemble mode
+    if isinstance(initial_state, list):
```
Comment (aaronleesander) [inline src/mqt/yaqs/simulator.py:920]: This is a good entry point

File: src/mqt/yaqs/simulator.py [hunk 7]
```diff
@@ -901,26 +1075,45 @@ def run(
...
+            selected simulation mode.
 
     """
     # Ensure the state is in B-normalization before any evolution
-    initial_state.normalize("B")
+    if isinstance(initial_state, list):
+        if any(not isinstance(state, MPS) for state in initial_state):
+            msg = "initial_state list must contain only MPS objects."
+            raise TypeError(msg)
+        typed_initial_states = cast("list[MPS]", initial_state)
```
Comment (aaronleesander) [inline src/mqt/yaqs/simulator.py]: I don't want to name things like "typed" since that should be a general part of the library. Can you make it like inital_state_list or initial_states or something else?

File: tests/analog/test_autocorrelator_periodic_wrap.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
+# Copyright (c) 2025 - 2026 Chair for Design Automation, TUM
```
Comment (aaronleesander) [inline tests/analog/test_autocorrelator_periodic_wrap.py]: For all the tests, instead of having separate test files like this, can you follow the exact structure of the source code? So for example, tests/analog/test_unitary_ensemble.py is fine since the file unitary_ensemble exists there. This is a stylistic choice I made for the library and isn't necessarily standard or non-standard, but I would rather keep it clean that way

File: CHANGELOG.md [hunk 9]
```diff
@@ -11,6 +11,10 @@ This project adheres to [Semantic Versioning], with the exception that minor rel
 
 ### Added
 
+- added deterministic unitary analog ensemble evolution by allowing `simulator.run` to accept `list[MPS]` as `initial_state` ([#409])
+- added optional autocorrelator evaluation (`compute_autocorrelator`) for unitary analog ensemble runs, with results written to `autocorrelator_results` on the simulation time grid ([#409])
+- added optional two-time correlator evaluation (`two_time_correlators`) for unitary analog ensemble runs, with results written to `two_time_correlator_results` on the simulation time grid ([#409])
```
Comment (aaronleesander) [inline CHANGELOG.md]: Your name should be on these as well (see the other ones). Since this is one pull request, maybe condense it. Typically we do the same name as the PR itself.

File: UPGRADING.md [hunk 10]
```diff
@@ -4,6 +4,49 @@ This document describes breaking changes and how to upgrade. For a complete list
 
 ## [Unreleased]
 
+### Analog `simulator.run` now accepts `list[MPS]` for unitary ensembles
```
Comment (aaronleesander) [inline UPGRADING.md]: Woops, I just saw this from the last PR as well. We should remove both of these. UPGRADING.md is more structural and not based on features (like when we deprecate a Python version)

File: docs/examples/two_time_correlators_unitary_analog_simulation.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    observables=[Observable(Z(), mid)],
+    elapsed_time=5.0,
+    dt=0.15,
+    max_bond_dim=64,
+    threshold=1e-10,
+    sample_timesteps=True,
+    show_progress=False,
+)
+
+simulator.run(psi0, H_open, primer_params, noise_model=None, parallel=False)
```
Comment (aaronleesander) [inline docs/examples/two_time_correlators_unitary_analog_simulation.md]: Can you double check the example just to make sure it does what you want it to do? AI likes to set parallel=False for some reason even though this is unnecessary. Then you explicitly say noise_model=None, but in reality, you just need to not feed anything into the function at all.

**Review by aaronleesander — CHANGES_REQUESTED** [review]
@Gauthameshwar Looks good overall, most of the comments are about organization stuff

File: docs/examples/ensemble_evolution.md
Comment (Gauthameshwar) [reply docs/examples/ensemble_evolution.md:1]: I have chosen noiseless_ensemble_evolution as the filename. I couldn't think of a shorter name because I want to highlight that we are doing a unitary dynamics here. Hope it is not too long to be displayed in the doc links.

File: docs/examples/two_time_correlators_unitary_analog_simulation.md (hunk 11 above)
Comment (Gauthameshwar) [reply docs/examples/two_time_correlators_unitary_analog_simulation.md]: I kept parallel=False for the single MPS run because parallelisation offers no advantage here. However, I noticed that it is set to False everywhere below as well. I have changed it. Also, not inputting anything at all is a good call, because you have a default value set to None for noise_model anyway. I have written a line highlighting this in my updated commit (to be pushed). Also, I noticed that "correlations" and "correlators" are used interchangeably in the docs. I have fixed that as well.

File: docs/refs.bib
Comment (Gauthameshwar) [reply docs/refs.bib:1]: My bad. I thought this was the place to hold the bibliography for any paper you cite in this package.

File: src/mqt/yaqs/analog/autocorrelator.py (hunk 1 above)
Comment (Gauthameshwar) [reply src/mqt/yaqs/analog/autocorrelator.py]: I also feel it would fit inside networks.py as an MPS method, because it mutates an MPS locally. If you are fine with relocating this function, I can move it to networks.py as apply_local().

File: src/mqt/yaqs/analog/autocorrelator.py (hunk 2 above)
Comment (Gauthameshwar) [reply src/mqt/yaqs/analog/autocorrelator.py]: Good idea. Will do it. That removes the need to have autocorrelator.py entirely.

File: src/mqt/yaqs/analog/ensemble.py (hunk 3 above)
Comment (Gauthameshwar) [reply src/mqt/yaqs/analog/ensemble.py:1]: Got it. Made the changes.

File: UPGRADING.md (hunk 10 above)
Comment (Gauthameshwar) [reply UPGRADING.md]: Done

File: CHANGELOG.md (hunk 9 above)
Comment (Gauthameshwar) [reply CHANGELOG.md]: Done.

File: tests/analog/test_autocorrelator_periodic_wrap.py (hunk 8 above)
Comment (Gauthameshwar) [reply tests/analog/test_autocorrelator_periodic_wrap.py]: I have moved the tests for mixed expectation and pbc to test_networks.py. I have moved the remaining dense matrix comparisons to test_spin_chain_ensemble.py (renamed test_xxz_transverse_unitary_ensemble.py)

**Comment by Gauthameshwar** [comment]
I have mostly made all the changes you requested including the shorter naming of functions and files, relocating functions like `apply_local()` and `mixed_expectation()` to the `MPS` class, tidier test files, and more concise scripts. The only thing remaining is the option to have the `two_time_correlators` and `autocorrelators` as a kind of `Observable`. 

I think the two-time correlator operator is fundamentally different from an observable in an MPO sense. The way this package handles observables is _at a given time_, compute $\langle \psi(t) | O | \psi(t)\rangle$. This is different for a two-time correlator because we are connecting two points in time here, and observables usually deal with the Hilbert space at a snapshot in time (for example, it makes sense to call two-point correlator functions $C(\sigma^x_i, \sigma^x_j) = \langle \sigma^x_i \sigma^x_j \rangle - \langle \sigma^x_i \rangle \langle \sigma^x_j \rangle$ an observable in this package because it is obtained from a state at a specific time). In a two_time_correlator $\langle A(t)B(0)\rangle$, you have two branches of $|\psi(t)\rangle$, one with a $B$ action, and the other without, and their mixed expectation on $A$. This is a completely different measurement protocol. So I could instead, reduce the API of `AnalogSimParams` by doing one of the following: 

1. Have two different structs for autocorrelator and two-time correlator, and input then inside a single kwarg: `AnalogSimParams(
    observables=[Observable()],
    correlators=[
    AutoCorrelator(O),
    TwoTimeCorrelator(A, B),
]
)`
2. Remove `autocorrelator` and `compute_autocorrelator`/`compute_twotime_correlator` completely and have the user only define the tuple of operators they want to use for computing two-time correlation: 
`AnalogSimParams(
    observables=[Observable()],
    two_time_correlators=[(O, O) #autocorrelation, (A, B) #twotimecorrelation]
)`

Personally, option 2 looks neat and the least invasive. What are your thoughts? Do you have any other suggestions?

File: src/mqt/yaqs/analog/autocorrelator.py (hunk 1 above)
Comment (aaronleesander) [reply src/mqt/yaqs/analog/autocorrelator.py]: Sure let's do apply_local

File: docs/examples/ensemble_evolution.md
Comment (aaronleesander) [reply docs/examples/ensemble_evolution.md:1]: I would drop the noiseless and just do ensemble evolution. They can then read it and see there is no noise.

**Comment by aaronleesander** [comment]
> I have mostly made all the changes you requested including the shorter naming of functions and files, relocating functions like `apply_local()` and `mixed_expectation()` to the `MPS` class, tidier test files, and more concise scripts. The only thing remaining is the option to have the `two_time_correlators` and `autocorrelators` as a kind of `Observable`.
> 
> I think the two-time correlator operator is fundamentally different from an observable in an MPO sense. The way this package handles observables is _at a given time_, compute ⟨ ψ ( t ) | O | ψ ( t ) ⟩ . This is different for a two-time correlator because we are connecting two points in time here, and observables usually deal with the Hilbert space at a snapshot in time (for example, it makes sense to call two-point correlator functions C ( σ i x , σ j x ) = ⟨ σ i x σ j x ⟩ − ⟨ σ i x ⟩ ⟨ σ j x ⟩ an observable in this package because it is obtained from a state at a specific time). In a two_time_correlator ⟨ A ( t ) B ( 0 ) ⟩ , you have two branches of | ψ ( t ) ⟩ , one with a B action, and the other without, and their mixed expectation on A . This is a completely different measurement protocol. So I could instead, reduce the API of `AnalogSimParams` by doing one of the following:
> 
> 1. Have two different structs for autocorrelator and two-time correlator, and input then inside a single kwarg: `AnalogSimParams( observables=[Observable()], correlators=[ AutoCorrelator(O), TwoTimeCorrelator(A, B), ] )`
> 2. Remove `autocorrelator` and `compute_autocorrelator`/`compute_twotime_correlator` completely and have the user only define the tuple of operators they want to use for computing two-time correlation:
>    `AnalogSimParams( observables=[Observable()], two_time_correlators=[(O, O) #autocorrelation, (A, B) #twotimecorrelation] )`
> 
> Personally, option 2 looks neat and the least invasive. What are your thoughts? Do you have any other suggestions?

I think the second one looks best. How about even better we do multi_time_observable. Then it has a clear analogy to the observable argument (and it is also an observable anyway).

File: tests/analog/test_autocorrelator_periodic_wrap.py (hunk 8 above)
Comment (aaronleesander) [reply tests/analog/test_autocorrelator_periodic_wrap.py]: Just combine it into test_ensemble.py, again so that it just mimics the src structure

**Comment by aaronleesander** [comment]
By the way, sometimes tests will fail just from the stochastic nature of the tests. They are mostly stable but there is always still a chance especially when we test it over and over again.

**Comment by Gauthameshwar** [comment]
Here are the updated changes. I have removed `autocorrelator` and `two_time_correlation` kwargs support, and put them all under `multi_time_observables`. I personally feel multi-time observables can be a bit misleading, in the sense that it could also imply we can input more than 2 observables at different times and still call it a multi-time observable. Now we only support a tuple of 2 observables. But that does not break the meaning of whatever is offered by your package at this point. So I decided to keep it this way for now. If you have ideas to support more than two observables to obtain their multi-time correlations, that could be part of the future development. 

P.S: For the tests passing/failing because of the stochastic nature of your tests, you can try to have pseudo-randomisation incorporated in your test files so they reproduce the same "random" scenario at every run. It might keep things consistent across multiple CI runs.

**Comment by aaronleesander** [comment]
@coderabbitai full review

**Review by aaronleesander — APPROVED** [review]
